File: 62-unique-paths/62-unique-paths.cpp
/*
    Reference:
        https://leetcode.com/problems/unique-paths/discuss/1581998/C%2B%2BPython-5-Simple-Solutions
        -w-Explanation-or-Optimization-from-Brute-Force-to-DP-to-Math

*/
class Solution {
public:
    int uniquePaths(int m, int n) {
        vector<vector<int>> dp(m, vector<int>(n, 1));
        for (int i = 1; i < m; ++i) {
            for (int j = 1; j < n; ++j) {
                dp[i][j] = dp[i - 1][j] + dp[i][j - 1];
            }
        }
        return dp[m - 1][n - 1];
    }
};

File: 62-unique-paths/62-unique-paths.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        dp = [[-1 for _ in range(n)] for _ in range(m)]
        
        def solve(i, j):
            if i == 0 and j == 0:
                return 1
            if i < 0 or j < 0:
                return 0;
            if dp[i][j] is not -1:
                return dp[i][j]
            dp[i][j] = solve(i - 1, j) + solve(i, j - 1);
            return dp[i][j]
        
        return solve(m - 1, n - 1)